core/models.py

```diff
@@ -581,37 +581,52 @@ def save_history(self):
     def save(self, *args, **kwargs):
         #get the user data so as to assign later his uuid id in fields user_updated etc
         user = User.objects.get(**kwargs)
-        now = py_datetime.now()
+        from core import datetime
+        now = datetime.datetime.now()
         #check if object has been newly created
         if self.id is None:
-           #save the new object
-           self.set_pk()
-           self.user_created = user
-           self.user_updated = user
-           self.date_created = now
-           self.date_updated = now
-           return super(HistoryModel, self).save()
+            #save the new object
+            self.set_pk()
+            self.user_created = user
+            self.user_updated = user
+            self.date_created = now
+            self.date_updated = now
+            return super(HistoryModel, self).save()
         if self.is_dirty(check_relationship=True):
-           self.date_updated = now
-           self.user_updated = user
-           self.version = self.version + 1
-           return super(HistoryModel, self).save()
+            self.date_updated = now
+            self.user_updated = user
+            self.version = self.version + 1
+            # check if we have business model
+            if hasattr(self, "replacement_uuid"):
+                if self.replacement_uuid is not None and 'replacement_uuid' not in self.get_dirty_fields():
+                    raise ValidationError('Update error! You cannot update replaced entity')
+            return super(HistoryModel, self).save()
         else:
-           raise ValidationError('Record has not be updated - there are no changes in fields')
+            raise ValidationError('Record has not be updated - there are no changes in fields')
 
     def delete_history(self):
         pass
 
     def delete(self, *args, **kwargs):
         user = User.objects.get(**kwargs)
         if not self.is_dirty(check_relationship=True) and not self.is_deleted:
-           self.date_updated = py_datetime.now()
-           self.user_updated = user
-           self.version = self.version + 1
-           self.is_deleted = True
-           return super(HistoryModel, self).save()
+            from core import datetime
+            now = datetime.datetime.now()
+            self.date_updated = now
+            self.user_updated = user
+            self.version = self.version + 1
+            self.is_deleted = True
+            #check if we have business model
+            if hasattr(self, "replacement_uuid"):
```

Comment:
Changes related to the point 3 - When a replacement entity is deleted, the link should be removed from replaced entity so a new replacement could be generated